pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -663,6 +671,14 @@ class Utils {
         }
     }
 
+    // Utility method to determine is triggers in PipelineTriggersJobProperties are equal
+    static boolean areTriggersInPipelineTriggerJobPropertiesEqual(PipelineTriggersJobProperty left, PipelineTriggersJobProperty right) {
+        if (left == right) return true;
+        if (left == null || right == null) return false;
+        boolean triggersEqual = CollectionUtils.isEqualCollection(left.triggers, right.triggers);
```

Comment:
Is this comparison unordered?  Will it return true even if only the order of the triggers has changed?

Reply:
@bitwiseman , @scoheb , sorry for interrupting your review, but I think this will not work.
`CollectionUtils.isEqualCollection()` use hashCode for comparing objects, but Trigger and child classes don't override this method, so it will not work.

But the problem is more complex, than the problem with Triggers. 
The whole method `org.jenkinsci.plugins.pipeline.modeldefinitionUtils.updateJobProperties` should be refactored, because it isn't thread-safe (I can be one of the reasons of [JENKINS-44809](https://issues.jenkins-ci.org/browse/JENKINS-44809)) and affect performance in case of parallel pipeline execution (save configuration each time, even if it has not been changed).

I'm working on a solution, which will resolve performance problems.
I want to add some hints, based on specifics of tipis of JobProperties. 
But exactly for triggers, I can't find a solution better than comparing Strings after XSTREAM serialization. Maybe you can advise me something before PR?

Reply:
@gulyaev13 
Interesting. Given what you're saying with hash code I wonder how the test succeeded? 

As to multi-threading - that would mean the the current running implementation is also not thread-safe.  Is what you have seen and are trying to fix?

Reply:
@bitwiseman , there is no magic. TestTrigger defined in the test has overwritten methods hashCode() and equals().

Yes, the current implementation isn't seen thread-safe, because of method updateJobProperties change global object WorkflowJob without any sync. 
I've seen the effects of concurrent changing a few times ago. The job's parameters were deleted from one thread and were saved to disk by another thread.
The main problem which I'm trying to solve is reducing a lot of IO calls caused by saving same configuration each job run (Save action is thread-save and become a bottleneck of concurrent pipeline execution).

Reply:
@gulyaev13 @bitwiseman I think the PR still improves the situation for triggers that do implement `equals()` and `hashcode()`

Reply:
@scoheb, I think I'm ready to send my PR which will solve your problem, too!
Am I add your test to my PR?

Reply:
@gulyaev13 of course you can use my test! thanks! looking forward to your PR!